### PR TITLE
modernize C-Fortran interface for cread functions

### LINKED ISSUE #421: old-style function definitions used in cread.c should be converted to modern style
In cread.c we have code like this:

```
/**
 * This subroutine opens a new system file for reading BUFR messages.
 *
 * @param[in] nfile -- f77int*: Internal Fortran I/O stream index associated with ufile
 * @param[in] ufile -- char*: [path/]name of system file to be opened
 *
 * @author J. Woollen @date 2012-09-15
 */
void openrb   (nfile,ufile) f77int *nfile; char *ufile; { pb[*nfile] = fopen( ufile , "rb " ); }
```

Note the old-style function definition.

Instead, this should be:

```
void 
openrb(f77int *nfile, char *ufile)
{ 
   pb[*nfile] = fopen( ufile , "rb " ); 
}

```

Note that the modern style of function definition is used. It has benefits over the old-style (see https://splichal.eu/scripts/sphinx/gcc/_build/html/extensions-to-the-c-language-family/prototypes-and-old-style-function-definitions.html) but even if it did not, the modern style is more compact and familiar to C programmers.

Note also that putting the entire function on one line of code takes the terseness of C beyond its usual limits. Let's keep at least one line per semicolon, as is commonly done in C code.

However this must be done carefully. Old-style function definitions cause less argument checking to be done, so there may be some cases where incorrect number or type of arguments are being passed to these functions, and changing to modern style function definitions will cause something to break.

Testing should be complete before anything beyond whitespace formatting is attempted, and also a bunch of C-only tests for these functions so that we're not just testing them indirectly, but have direct testing of everything. See https://ee.hawaii.edu/~tep/EE160/Book/chapapx/node7.html for some detail about other pitfalls to watch out for.

### PR DESCRIPTION
**Pull request by jbathegit**
Fixes #421

Part of #79
Part of #254

### CODE REVIEW

**Comment by jbathegit** [comment]
Hmm, so this is failing in Doxygen in the developer CI.  It took me a while to find it, but the applicable message is:
```
Generating docs for namespace bufrlib
Generating docs for compound bufrlib::backbufr_c...
Exiting...
make[2]: *** [docs/CMakeFiles/doc.dir/build.make:71: docs/CMakeFiles/doc] Error 1
make[1]: *** [CMakeFiles/Makefile2:2377: docs/CMakeFiles/doc.dir/all] Error 2
```
which is an interface defined in the new `bufrlib.F90` module.  I tried to model this on the earlier code in `bufr_interface.F90` where Doxygen was being squawky about the `@fn` notation, and (I think) the only difference here is that this module name is bufrlib instead of bufr_interface, but I seem to be missing something else.  @edwardhartnett, any thoughts?

File: src/bufr_interface.F90
Comment (edwardhartnett) [inline src/bufr_interface.F90]: Two points:
1. We don't need to say "NCEPLIBS-bufr ccbfl() function", we can just say "Wraps ccbfl()."
2. This is not a good brief because the user will not know what the ccbfl() function does. Instead, these functions should have briefs like the functions they wrap, and then the fact that they are wrapping should be mentioned but not in the brief. Something like:

    !> @fn bufr_interface::ccbfl_c::ccbfl_c()
    !> Closes all system files that were opened via previous calls to subroutine cobfl().
    !>
    !> Wraps ccbfl().

File: src/bufr_interface.h
Comment (edwardhartnett) [inline src/bufr_interface.h]: Usually we say "Fortran logical unit number" not "Fortran file unit number".

File: src/bufrlib.F90
Comment (edwardhartnett) [inline src/bufrlib.F90]: Take out the first three words so this makes a good brief description of the module.

File: src/bufrlib.h.in
Comment (edwardhartnett) [inline src/bufrlib.h.in]: "a number of NCEPLIBS-bufr subprograms?" Can we be more specific?

File: src/cpdxmm.f
Comment (edwardhartnett) [inline src/cpdxmm.f:63]: I don't think you need to have _C at the end of the function names, and if you didn't then this code would not have to change.

File: src/cread.c
Comment (edwardhartnett) [inline src/cread.c]: The way I would normally phrase this is:

@param mxwd - number of elements in array bufr.

This makes more sense in C, where generally we don't dimension large arrays, but instead allocate them at runtime. "Number of elements" makes sense whether array is allocated or dimensions.

File: src/cread.c
Comment (edwardhartnett) [inline src/cread.c:106]: Thanks for breaking this up. It looks like something from the infamous obfuscated C contest...

File: src/crwbmg.c
Comment (edwardhartnett) [inline src/crwbmg.c]: Let's change "system file" to just "file".

File: src/openbf.f
Comment (edwardhartnett) [inline src/openbf.f:13]: Remove "typically". Is there actually any other way to associate a logical unit with a file in Fortran?

File: src/rdbfdx.f
Comment (edwardhartnett) [inline src/rdbfdx.f:29]: Can you explain why we need the use statement now? We didn't need it before...

File: src/rdmsgw.F90
Comment (edwardhartnett) [inline src/rdmsgw.F90]: Can this be coded without the goto?

File: src/bufr_interface.h
Comment (jbathegit) [reply src/bufr_interface.h]: This was how Ron had originally worded it, but I see your point.  I'll get this changed.

File: src/bufrlib.h.in
Comment (jbathegit) [reply src/bufrlib.h.in]: We could, but it would be a long list of Fortran and C routines.

File: src/cpdxmm.f
Comment (jbathegit) [reply src/cpdxmm.f:63]: You may be right.  I was just following Ron's earlier approach that he used in all of the bufr_c2f_interface routines, where anytime a C code was being called from Fortran it had a _C suffix, and where anytime a Fortran code was being called from C it had a _F suffix.  You'll see this in a number of test codes as well, and I think it does provide a nice intuitive indicator/reminder that the code being called is a wrapper to a different language.  And it helps to self-document the calling code as well.

File: src/cread.c
Comment (jbathegit) [reply src/cread.c]: OK.  I'll change this in crwbmg.c as well.

Comment (jbathegit) [reply src/cread.c]: And in srchtbf.c too :-)

File: src/crwbmg.c
Comment (jbathegit) [reply src/crwbmg.c]: OK, good catch.

Comment (jbathegit) [reply src/crwbmg.c]: I'll go ahead now and remove all of the "system file" references throughout the library to be just "file".

File: src/openbf.f
Comment (jbathegit) [reply src/openbf.f:13]: Actually there is.  Some operating systems let you associate a file with Fortran logical unit X by setting FORT_X=file within the UNIX script which calls the code, rather than using an OPEN statement inside the code.

File: src/rdbfdx.f
Comment (jbathegit) [reply src/rdbfdx.f:29]: We're trying to modernize the C-Fortran interface and get rid of the UNDERSCORE and other legacy stuff, so we now have wrappers with `BIND(C,` defined in a module interface in bufrlib.F90.  So that's why we need to explicitly include the new module whenever we call these C routines from Fortran inside of the library.

This is similar to how all of the test codes now need to `use bufr_interface` when calling C from Fortran, or `#include bufr_interface.h` when calling Fortran from C.

File: src/bufr_interface.F90
Comment (jbathegit) [reply src/bufr_interface.F90]: 1. OK, fair point.
2. OK, but this seems redundant, because we're already noting that this is a wrapper for a different function, so the user could always just go to that referenced function to see what it does.

File: src/cread.c
Comment (edwardhartnett) [inline src/cread.c]: Frequently in C there is a convention to define the function like this (note the function type is on a line by itself and the function name starts in the very first column of its own line):

```
void 
openrb( int nfile, char *ufile ) {

```

This is done only where the function is defined, not for any prototype. The reason is to make the function easy to find. Here's an example that just came up from netcdf-c code, to illustrate.

I was looking for the function nc_set_log_level() in the netcdf-c code, so I did this:


```
ed@koko:~/netcdf-c$ find . -name '*.c'|xargs grep nc_set_log_level
./nc_test4/tst_rename3.c:      nc_set_log_level(4);
./nc_test4/tst_rename3.c:      nc_set_log_level(4);
./nc_test4/tst_rename3.c:      nc_set_log_level(4);
./nc_test4/tst_vars4.c:        nc_set_log_level(-1);
./nc_test4/tst_vars4.c:        nc_set_log_level(-1);
./nc_test4/test_filter_order.c:    nc_set_log_level(1);
./nc_test4/tst_fills.c:      nc_set_log_level(0);
./nc_test4/tst_rename.c:   /* nc_set_log_level(5); */
./nc_test4/tst_rename.c:         nc_set_log_level(5);
./nc_test4/test_filterx_hdf5.c:    nc_set_log_level(1);
./nc_test4/tst_parallel_compress.c:                /* nc_set_log_level(3); */
./nc_test4/tst_parallel_compress.c:                /* nc_set_log_level(3); */
./nc_test4/test_filter_repeat.c:    nc_set_log_level(1);
./nc_test4/test_filter_misc.c:    nc_set_log_level(1);
./nc_test4/test_filter_vlen.c:    nc_set_log_level(1);
./nc_test4/tst_interops5.c:/*	 nc_set_log_level(5);*/
./nc_test4/tst_interops5.c:       nc_set_log_level(3);
./nc_test4/tst_h_refs.c:        nc_set_log_level(4);
./nc_test4/test_filter_reg.c:    nc_set_log_level(1);
./nc_test4/tst_dims2.c:      nc_set_log_level(0);
./nc_test4/tst_parallel6.c:    /* nc_set_log_level(4); */
./nc_test4/tst_parallel6.c:    /* nc_set_log_level(-1); */
./nc_test4/tst_v2.c:   /*nc_set_log_level(3);*/
./nc_test4/tst_filter_avail.c:    nc_set_log_level(1);
./nc_test4/tst_rename2.c:      /* nc_set_log_level(4); */
./nc_test4/tst_rename2.c:      nc_set_log_level(4);
./nc_test4/tst_rename2.c:      nc_set_log_level(4);
./nc_test4/tst_parallel.c:    /*nc_set_log_level(3);*/
./nc_test4/tst_parallel3.c:/*    nc_set_log_level(NC_TURN_OFF_LOGGING); */
./nc_test4/tst_parallel3.c:   /*    nc_set_log_level(3);*/
./nc_test4/tst_parallel3.c:/*      nc_set_log_level(NC_TURN_OFF_LOGGING); */
./nc_test4/tst_parallel3.c:/*      nc_set_log_level(4);*/
./nc_test4/tst_parallel5.c:            /* nc_set_log_level(3); */
./nc_test4/tst_parallel5.c:            /* nc_set_log_level(-1); */
./nc_test4/tst_parallel_zlib.c:    /*nc_set_log_level(3);*/
./nc_test4/tst_parallel_zlib.c:       nc_set_log_level(3);
./nc_test4/renamegroup.c:    stat = nc_set_log_level(0);
./nc_test4/tst_enums.c:      /*nc_set_log_level(3);*/
./nc_test4/tst_varms.c:      /*nc_set_log_level(2);*/
./libsrc4/nc4internal.c:nc_set_log_level(int new_level)
./libsrc4/nc4dispatch.c:            nc_set_log_level((int)level);
./libnczarr/zinternal.c: * @internal This is like nc_set_log_level(), but will also turn on
./nc_test/nc_test.c:    /* If you uncomment the nc_set_log_level line, you will get a lot
./nc_test/nc_test.c:    /*nc_set_log_level(3);*/
./nc_test/tst_small.c:   /*nc_set_log_level(3);*/
./ncgen/genbin.c:    nc_set_log_level(ncloglevel);
./ncgen/genc.c:        bbprintf0(stmt,"    nc_set_log_level(%d); /* set log level */",ncloglevel);
./ncdump/tst_create_files.c:   /*nc_set_log_level(4);*/
./ncdump/nccopy.c:              nc_set_log_level(level);
./ncdump/tst_enum_data.c:   /*nc_set_log_level(3);*/
./ncdump/ncdump.c:	    nc_set_log_level(level);
./ncdump/tst_enum_undef.c:   /*nc_set_log_level(3);*/
./libhdf5/hdf5internal.c: * @internal This is like nc_set_log_level(), but will also turn on
./nczarr_test/testfilter_order.c:    nc_set_log_level(1);
./nczarr_test/test_filter_order.c:    nc_set_log_level(1);
./nczarr_test/test_filter_repeat.c:    nc_set_log_level(1);
./nczarr_test/test_filter_misc.c:    nc_set_log_level(1);
./nczarr_test/testfilter_misc.c:    nc_set_log_level(1);
./nczarr_test/test_filter_avail.c:    nc_set_log_level(1);
./nczarr_test/testfilter_repeat.c:    nc_set_log_level(1);
./nc_perf/tst_bm_rando.c:        /* nc_set_log_level(3); */
./nc_perf/tst_compress.c:                        /* nc_set_log_level(3); */
./nc_perf/tst_compress_par.c:		    /* nc_set_log_level(3); */
./nc_perf/tst_wrf_reads.c:               nc_set_log_level(4);
./nc_perf/tst_wrf_reads.c:   /*             nc_set_log_level(4); */
./nc_perf/tst_wrf_reads.c:   /*             nc_set_log_level(4); */
./b/nczarr_test/test_filter_order.c:    nc_set_log_level(1);
./b/nczarr_test/test_filter_repeat.c:    nc_set_log_level(1);
./b/nczarr_test/test_filter_misc.c:    nc_set_log_level(1);
./b/nczarr_test/test_filter_avail.c:    nc_set_log_level(1);
```
Too many results! This function is used everywhere in the code base.

Then I remembered the convention and grep'd with the ^ in front of the function name:
```
ed@koko:~/netcdf-c$ find . -name '*.c'|xargs grep ^nc_set_log_level
./libsrc4/nc4internal.c:nc_set_log_level(int new_level)

```

Now I see where my function is. I suggest you adopt this convention as well. It's common in the C world.

Also, while a space before and after parens may be OK in Fortran, I've never seen anyone do it in C. Perhaps that's a custom best left in the Fortran code.

There is strong project management value in having code that looks like everyone else's code. It makes it easier for other programmers, including outside collaborators, to more easily understand the code.

NetCDF for example has become so much better due to years of contributions of power-users. This project can benefit in a similar way, but we need to eliminate barriers to entry and make our code attractive and easy to work with. Or work in documentation and testing really help in this regard.

File: src/bufrlib.F90
Comment (jbathegit) [reply src/bufrlib.F90]: OK, done.

File: src/rdmsgw.F90
Comment (jbathegit) [reply src/rdmsgw.F90]: Done.

File: src/cread.c
Comment (jbathegit) [reply src/cread.c]: re: anchoring the function to the first column, I wasn't aware that was a common convention.  I don't mind doing that, but you'll have to give me a few minutes to push up another commit, because there's quite a bit of C code in the library.

re: the spacing before and after the parentheses, I like them for readability.  But you probably have more experience in the C world than me, so I'll defer to your judgement and if you think that's also an oddball feature then I'll take those out too.

**Comment by jbathegit** [comment]
In the meantime, could you please help me figure out why Doxygen is squawking in the developer CI build? ;-)  That was my original problem, and I can't reproduce it with the older version of Doxygen on my local machine.  I honestly don't understand what it's complaining about.

File: src/bufr_interface.F90
Comment (edwardhartnett) [reply src/bufr_interface.F90]: But when the user hovers above the function name, he or she sees "Wraps gobbledy-goop" which is not particularly helpful.

The brief needs to tell what the function does in a way the user quickly understands, so he can scan the functions listed in the docs as say "ah HA! THAT'S the one I want!"

File: src/bufrlib.h.in
Comment (edwardhartnett) [reply src/bufrlib.h.in]: And do they have nothing in common other than the fact they are in NCEPLIBS-bufr? Were they selected randomly?

File: src/cpdxmm.f
Comment (edwardhartnett) [reply src/cpdxmm.f:63]: I don't object. I just wanted to be sure I understood what was happening.

File: src/openbf.f
Comment (edwardhartnett) [reply src/openbf.f:13]: And do we support that? (That is, do we test it?)

File: src/rdbfdx.f
Comment (edwardhartnett) [reply src/rdbfdx.f:29]: OK, that makes sense. SOrry, I'm a little slow sometimes! ;-)

**Comment by edwardhartnett** [comment]
Yes I will take a look at doxygen. Stand by...

File: docs/ReleaseNotes.md
Comment (edwardhartnett) [inline docs/ReleaseNotes.md]: Here is your first doxygen problem. It generates this error:

/home/ed/NCEPLIBS-bufr/docs/ReleaseNotes.md:24: error: explicit link request to 'include' could not be resolved (warning treated as error, aborting now)

The problem is you have a hash sign before the include, so doxygen is thinking this is a doxygen command and not being able to cope.

When you want to use a hash in text, you need to escape it with a backslash, like this:

<pre><b>\#include "bufr_interface.h"</b> statement to directly call any Fortran</pre>

Comment (edwardhartnett) [reply docs/ReleaseNotes.md]: After that is resolved you have a bunch of doxygen warnings about functions without all the parameters defined. Let me know if you need help with those.

One thing that makes debugging from the CI hard is the make -j command. When we use -j with make we get faster builds, but it jumbles together the output from different processes. So when confused, sometimes I take the -j out of the make command in the developer workflow, to get more clear output. Just remember to put it back in, because normally we want the faster compiles.

File: src/bufr_interface.F90
Comment (jbathegit) [reply src/bufr_interface.F90]: OK, got it.  I've already done this in bufr_interface.F90 and bufrlib.F90, BTW.  But I'll go ahead and add those now in bufr_c2f_interface.F90 as well, which never had them.

File: docs/ReleaseNotes.md
Comment (jbathegit) [reply docs/ReleaseNotes.md]: OK thanks.  Can you please let me know what functions don't have all the parameters defined?

File: src/bufrlib.h.in
Comment (jbathegit) [reply src/bufrlib.h.in]: What they have in common is that they're all called from within C code inside of the library.  Some of them are native C so they only need prototypes defined, while others are native Fortran so they need wrappers defined.

File: docs/ReleaseNotes.md
Comment (jbathegit) [reply docs/ReleaseNotes.md]: The backslash escape has been added, and I've also cleaned up the bufr_c2f_interface documentation as noted above.

But Doxygen is still not happy.  And again, I'm unable to reproduce this error on my system, even if I temporarily remove the -j2 switch.

**Comment by edwardhartnett** [comment]
How come you can't run doxygen?

```
Generating docs for compound bufrlib::backbufr_c...
/home/ed/NCEPLIBS-bufr/src/bufrlib.F90:57: warning: parameters of member bufrlib::backbufr_c::backbufr_c are not (all) documented
Generating docs for compound bufrlib::cewind_c...
/home/ed/NCEPLIBS-bufr/src/bufrlib.F90:68: warning: parameters of member bufrlib::cewind_c::cewind_c are not (all) documented
Generating docs for compound bufrlib::closfb_c...
/home/ed/NCEPLIBS-bufr/src/bufrlib.F90:79: warning: parameters of member bufrlib::closfb_c::closfb_c are not (all) documented
Generating docs for compound bufrlib::crdbufr_c...
/home/ed/NCEPLIBS-bufr/src/bufrlib.F90:90: warning: parameters of member bufrlib::crdbufr_c::crdbufr_c are not (all) documented
Generating docs for compound bufrlib::cwrbufr_c...
/home/ed/NCEPLIBS-bufr/src/bufrlib.F90:102: warning: parameters of member bufrlib::cwrbufr_c::cwrbufr_c are not (all) documented
Generating docs for compound bufrlib::openab_c...
/home/ed/NCEPLIBS-bufr/src/bufrlib.F90:45: warning: parameters of member bufrlib::openab_c::openab_c are not (all) documented
Generating docs for compound bufrlib::openrb_c...
/home/ed/NCEPLIBS-bufr/src/bufrlib.F90:21: warning: parameters of member bufrlib::openrb_c::openrb_c are not (all) documented
Generating docs for compound bufrlib::openwb_c...
/home/ed/NCEPLIBS-bufr/src/bufrlib.F90:33: warning: parameters of member bufrlib::openwb_c::openwb_c are not (all) documented

```

**Comment by edwardhartnett** [comment]
OK, I did a commit to this branch which turns off parallel builds in developer.yml. 

I changed this:

`        make -j2 VERBOSE=1
`
to:

`        make VERBOSE=1
`

Now we can clearly see the first doxyen error in the output:

`/home/runner/work/NCEPLIBS-bufr/NCEPLIBS-bufr/bufr/src/bufrlib.F90:57: error: parameters of member bufrlib::backbufr_c::backbufr_c are not (all) documented (warning treated as error, aborting now)
`

Since WARN_AS_ERROR is YES, in docs/Doxyfile.in, the build will break at the first documentation error. If you temporarily turn off WARN_AS_ERROR in your branch, the documentation build will succeed and you will see all the warnings. I will demonstrate...

**Comment by edwardhartnett** [comment]
OK, I turned off WARN_AS_ERROR in the Doxyfile.in. Now the developer workflow completes. Since I've also turned off parallel builds, it's easy to find all the doxygen warnings in the output:

```
/home/runner/work/NCEPLIBS-bufr/NCEPLIBS-bufr/bufr/src/bufrlib.F90:57: warning: parameters of member bufrlib::backbufr_c::backbufr_c are not (all) documented
/home/runner/work/NCEPLIBS-bufr/NCEPLIBS-bufr/bufr/src/bufrlib.F90:68: warning: parameters of member bufrlib::cewind_c::cewind_c are not (all) documented
/home/runner/work/NCEPLIBS-bufr/NCEPLIBS-bufr/bufr/src/bufrlib.F90:79: warning: parameters of member bufrlib::closfb_c::closfb_c are not (all) documented
/home/runner/work/NCEPLIBS-bufr/NCEPLIBS-bufr/bufr/src/bufrlib.F90:90: warning: parameters of member bufrlib::crdbufr_c::crdbufr_c are not (all) documented
/home/runner/work/NCEPLIBS-bufr/NCEPLIBS-bufr/bufr/src/bufrlib.F90:102: warning: parameters of member bufrlib::cwrbufr_c::cwrbufr_c are not (all) documented
/home/runner/work/NCEPLIBS-bufr/NCEPLIBS-bufr/bufr/src/bufrlib.F90:45: warning: parameters of member bufrlib::openab_c::openab_c are not (all) documented
/home/runner/work/NCEPLIBS-bufr/NCEPLIBS-bufr/bufr/src/bufrlib.F90:21: warning: parameters of member bufrlib::openrb_c::openrb_c are not (all) documented
/home/runner/work/NCEPLIBS-bufr/NCEPLIBS-bufr/bufr/src/bufrlib.F90:33: warning: parameters of member bufrlib::openwb_c::openwb_c are not (all) documented

```

So this is a demonstration of how you can easily see your doxygen warnings in the CI system,

But that's not how I do things. Instead, I have installed doxygen-1.9.1 on my machine and set my PATH to find that version first. Then I can run make in the build/docs directory after running cmake, and I get a doxygen build. I find that a little easier than waiting for the CI.

But the CI will work as well, so if you can't install doxygen on your system (but why can't you?) then you can use the CI instead.

**Comment by edwardhartnett** [comment]
Well you have not defined the params for lots of the functions. For example:

```
    !> @fn bufrlib::backbufr_c::backbufr_c(nfile)
    !> Backspace a BUFR file by one BUFR message.
    !>
    !> Wraps backbufr() function.
    !>
    !> @author J. Ator @date 2005-11-29
    subroutine backbufr_c( nfile ) bind(C, name='backbufr')

```
Where's the @param nfile line in the documentation?

**Comment by edwardhartnett** [comment]
OK, I'm going to leave parallel builds off, and WARN_AS_ERROR off, in your branch. Once you fix your warnings, make sure you turn them on again.

**Comment by jbathegit** [comment]
ARGGH, I can't believe I missed that!  Thanks for your help, and I'll get that fixed and pushed up.

As for why I can't install doxygen on my system, I'm working on WCOSS2, which has a strict rule about users installing their own 3rd-party libraries.  The system does have doxygen already installed, but it's a much older version 1.8.14, and it wasn't showing any of the above output when I ran it myself with `-DENABLE_DOCS=On`.  So thanks again for your help here!

**Comment by edwardhartnett** [comment]
OK, glad to help! Lot's of really great test improvements happening - it's encouraging to see. Last I checked, 82.4%. That number keeps climbing! ;-)

**Comment by jbathegit** [comment]
Thanks, and we should be good to go with this PR now.  If you could please approve it then I'll get it merged.  Or if you still have any concerns that I haven't addressed yet, please let me know.

**Review by edwardhartnett — APPROVED** [review]

**Review by jack-woollen — COMMENTED** [review]
Seems to me subroutine "bufr_interface" should be named "bufr_f2c_interface", and subroutine "bufrlib" should be named "cread_f2c_interface", for both the .F90 and .h parts. That makes the language interface nomenclature more consistent in the library.

**Pull request merged**
